substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/DwarfLocSectionImpl.java

```diff
@@ -0,0 +1,718 @@
+        if (buffer == null) {
+            pos += putByte(op, scratch, 0);
+            pos += putULEB(byteCount, scratch, 0);
+            if (constant.isNull()) {
+                pos += writeAttrData8(0, scratch, 0);
+            } else if (byteCount == 1) {
+                if (kind == JavaKind.Int) {
+                    pos += putByte((byte) constant.asInt(), scratch, 0);
+                } else {
+                    pos += putByte((byte) (constant.asBoolean() ? 1 : 0), scratch, 0);
+                }
+            } else if (byteCount == 2) {
+                pos += putShort((short) constant.asInt(), scratch, 0);
+            } else if (byteCount == 4) {
+                int i = (kind == JavaKind.Int ? constant.asInt() : Float.floatToRawIntBits(constant.asFloat()));
+                pos += putInt(i, scratch, 0);
+            } else {
+                long l = (kind == JavaKind.Long ? constant.asLong() : Double.doubleToRawLongBits(constant.asDouble()));
+                pos += putLong(l, scratch, 0);
+            }
+        } else {
+            pos = putByte(op, buffer, pos);
+            pos = putULEB(byteCount, buffer, pos);
+            if (constant.isNull()) {
+                pos = writeAttrData8(0, buffer, pos);
+            } else if (byteCount == 1) {
+                if (kind == JavaKind.Int) {
+                    pos = putByte((byte) constant.asInt(), buffer, pos);
+                } else {
+                    pos = putByte((byte) (constant.asBoolean() ? 1 : 0), buffer, pos);
+                }
+            } else if (byteCount == 2) {
+                pos = putShort((short) constant.asInt(), buffer, pos);
+            } else if (byteCount == 4) {
+                int i = (kind == JavaKind.Int ? constant.asInt() : Float.floatToRawIntBits(constant.asFloat()));
+                pos = putInt(i, buffer, pos);
+            } else {
+                long l = (kind == JavaKind.Long ? constant.asLong() : Double.doubleToRawLongBits(constant.asDouble()));
+                pos = putLong(l, buffer, pos);
+            }
+            verboseLog(context, "  [0x%08x]     CONSTANT %s", pos, constant.toValueString());
+        }
+        return pos;
+    }
+
+    // auxiliary class used to collect per-range locations for a given local
+    // merging adjacent ranges with the same location
+    static class LocalValueExtent {
+        long lo;
+        long hi;
+        DebugLocalValueInfo value;
+
+        LocalValueExtent(long lo, long hi, DebugLocalValueInfo value) {
+            this.lo = lo;
+            this.hi = hi;
+            this.value = value;
+        }
+
+        @SuppressWarnings("unused")
+        boolean shouldMerge(int otherLo, int otherHi, DebugLocalValueInfo otherValue) {
+            // ranges need to be contiguous to merge
+            if (hi != otherLo) {
+                return false;
+            }
+            // values need to be for the same line
+            if (value.line() != otherValue.line()) {
```

Comment:
I'm curious, does it really happen that when we have `hi == otherLo` that `value.line() != otherValue.line()` is ever true?

Reply:
Yes, it can happen. The line number here is not for the declaration of the local var but for the infopoint where a location was recorded for the local. So, we can have two successive infopoints with different line numbers that both provide values for the local.

However, your question raises a good point all the same. It is important not to merge Range instances with different lines because the Range objects are used to drive generation of debug line info. However, I am now wondering whether we need to care about the line difference when merging location extents. I don't think the debugger lookup for locations requires that the location ranges conform to the line number ranges. So, I think we can remove this check and rely on the subsequent value comparison to decide whether or not to merge  the location extents.

Reply:
This got folded into the equals method where we do have to test for equality of line numbers.